zrml/neo-swaps/src/math.rs

```diff
@@ -0,0 +1,282 @@
+            to_fixed(liquidity)?,
+        )?;
+        from_fixed(result_fixed)
+    }
+
+    /// Calculate e^(-r_i/b).
+    pub(super) fn calculate_spot_price(reserve: u128, liquidity: u128) -> Option<u128> {
+        let result_fixed = calculate_spot_price_fixed(to_fixed(reserve)?, to_fixed(liquidity)?)?;
+        from_fixed(result_fixed)
+    }
+
+    pub(super) fn calculate_reserves_from_spot_prices(
+        amount: u128,
+        spot_prices: Vec<u128>,
+    ) -> Option<(u128, Vec<u128>)> {
+        let (liquidity_fixed, reserve_fixed) = calculate_reserve_from_spot_prices_fixed(
+            to_fixed(amount)?,
+            spot_prices.into_iter().map(to_fixed).collect::<Option<Vec<_>>>()?,
+        )?;
+        let liquidity = from_fixed(liquidity_fixed)?;
+        let reserve = reserve_fixed.into_iter().map(from_fixed).collect::<Option<Vec<_>>>()?;
+        Some((liquidity, reserve))
+    }
+
+    fn to_fixed<B>(value: B) -> Option<Fixed>
+    where
+        B: Into<u128> + From<u128>,
+    {
+        value.to_fixed_from_fixed_decimal(DECIMALS).ok()
+    }
+
+    fn from_fixed<B>(value: Fixed) -> Option<B>
+    where
+        B: Into<u128> + From<u128>,
+    {
+        value.to_fixed_decimal(DECIMALS).ok()
+    }
+
+    fn calculate_swap_amount_out_for_buy_fixed(
+        reserve: Fixed,
+        amount_in: Fixed,
+        liquidity: Fixed,
+    ) -> Option<Fixed> {
+        // FIXME Defensive programming: Check for underflow in x/b and r_i/b.
+        let exp_x_over_b: Fixed = exp(amount_in.checked_div(liquidity)?, false).ok()?;
+        let exp_neg_r_over_b = exp(reserve.checked_div(liquidity)?, true).ok()?;
+        // FIXME Defensive programming: Check for underflow in the exponential expressions.
+        let inside_ln =
+            exp_x_over_b.checked_add(exp_neg_r_over_b)?.checked_sub(Fixed::checked_from_num(1)?)?;
+        let (ln_result, ln_neg) = ln(inside_ln).ok()?;
+        let blob = liquidity.checked_mul(ln_result)?;
+        let reserve_plus_blob =
+            if ln_neg { reserve.checked_sub(blob)? } else { reserve.checked_add(blob)? };
+        reserve_plus_blob.checked_sub(amount_in)
+    }
+
+    fn calculate_swap_amount_out_for_sell_fixed(
+        reserve: Fixed,
+        amount_in: Fixed,
+        liquidity: Fixed,
+    ) -> Option<Fixed> {
+        // FIXME Defensive programming: Check for underflow in x/b and r_i/b.
+        let exp_neg_x_over_b: Fixed = exp(amount_in.checked_div(liquidity)?, true).ok()?;
+        let exp_r_over_b = exp(reserve.checked_div(liquidity)?, false).ok()?;
+        // FIXME Defensive programming: Check for underflow in the exponential expressions.
+        let inside_ln =
+            exp_neg_x_over_b.checked_add(exp_r_over_b)?.checked_sub(Fixed::checked_from_num(1)?)?;
+        let (ln_result, ln_neg) = ln(inside_ln).ok()?;
+        let blob = liquidity.checked_mul(ln_result)?;
+        if ln_neg { reserve.checked_add(blob) } else { reserve.checked_sub(blob) }
+    }
```

Comment:
By the way, the same question is applicable here. But here it is inverted.

Reply:
Same argument applies, except that here the `b * ln(...)` term has a negative sign: `–1 * b * ln( e^(-x/b) − 1 + e^(r_i/b) ) + r_i`.